File: JavaScript/314-Binary Tree Vertical Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalOrder = function(root) {
    if (!root) return [];
    let queue = [[root, 0]];
    let res = new Map();
    let minCol = 0, maxCol = 0;
    while (queue.length){
        queueSize = queue.length;
        for (let i=0; i<queueSize; i++){
            let [node, pos] = queue.shift();
            res.set(pos, (res.get(pos)|| []).concat([node.val]) )
            if (node.left){
                queue.push([node.left, pos-1]);
                minCol = Math.min(minCol, pos-1);
            }
            if (node.right){
                queue.push([node.right, pos+1]);
                maxCol = Math.max(maxCol, pos+1);
            }
        }
    }
    let ans = [];
    for (let i=minCol; i<=maxCol; i++){
        if (res.has(i)){
            ans.push(res.get(i));
        }
    }
    return ans;
};

File: C++/314-Binary Tree Vertical Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        if (root == nullptr) {
            return {};
        }
        map<int, vector<int>> hmap;
        deque<pair<TreeNode*, int>> dq;

        dq.push_back({root, 0});

        while (!dq.empty()) {
            int sz = dq.size();
            while (sz-- > 0) {
                auto [node, idx] = dq.front();
                dq.pop_front();
                hmap[idx].push_back(node->val);
                if (node->left != nullptr) {
                    dq.push_back({node->left, idx-1});
                }
                if (node->right != nullptr) {
                    dq.push_back({node->right, idx+1});
                }
            }
        }

        vector<vector<int>> res;
        for (auto &[_, v]:hmap) {
            res.push_back(v);
        }

        return res;
    }
};